azure/services/virtualmachines/client.go

```diff
@@ -78,12 +78,25 @@ func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.Resou
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "virtualmachines.AzureClient.CreateOrUpdate")
 	defer done()
 
-	vm, err := ac.vmParams(ctx, spec)
+	var existingVM compute.VirtualMachine
+	var err error
+
+	if existingVM, err = ac.Get(ctx, spec.ResourceGroupName(), spec.ResourceName()); err != nil && !azure.ResourceNotFound(err) {
+		return nil, nil, errors.Wrapf(err, "failed to get virtual machine %s in %s", spec.ResourceName(), spec.ResourceGroupName())
+	}
+
+	params, err := spec.Parameters(existingVM)
 	if err != nil {
 		return nil, nil, errors.Wrapf(err, "failed to get desired parameters for virtual machine %s", spec.ResourceName())
-	} else if vm == nil {
-		// nothing to do here.
-		return nil, nil, nil
+	}
+
+	vm, ok := params.(compute.VirtualMachine)
+	if !ok {
+		if params == nil {
+			// nothing to do here.
+			return existingVM, nil, nil
```

Comment:
this addresses https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1697#discussion_r742238447

Reply:
can we move this to line 86 and invoke spec.Parameters() can be only for new vms? wdyt?

Reply:
in this specific VM service we technically could because we never want to update existing VMs but this pattern works best for all services in general: we pass in the existing resource to `spec.Parameters(existing)` and that function determines if the resource needs to be created or updated based on the existing resource. For example, for NSGs we update the group the security rules are missing. IMO it's best to use the same pattern throughout all services so it's easier to refactor them later to have even more code in common.